devito/types/grid.py

```diff
@@ -531,50 +533,112 @@ def __init__(self, **kwargs):
             n = Dimension(name='n')
             self.implicit_dimension = n
         self._n_domains = kwargs.get('N', 1)
-        self._bounds = kwargs.get('bounds', None)
+        self._global_bounds = kwargs.get('bounds', None)
+        self._implicit_exprs = None
 
-    def __subdomain_finalize__(self, dimensions, shape, **kwargs):
+    def __subdomain_finalize__(self, dimensions, shape, distributor=None, **kwargs):
         # Create the SubDomain's SubDimensions
         sub_dimensions = []
         for d in dimensions:
-            sub_dimensions.append(
-                SubDimension.middle('%si_%s' % (d.name, self.implicit_dimension.name),
-                                    d, 0, 0))
+            sub_dimensions.append(SubDimension.middle
+                                  ('%si_%s' % (d.name, self.implicit_dimension.name),
+                                   d, 0, 0))
         self._dimensions = tuple(sub_dimensions)
-        # Compute the SubDomain shape
-        self._shape = tuple(s - (sum(d._thickness_map.values()) if d.is_Sub else 0)
-                            for d, s in zip(self._dimensions, shape))
+
+        # Compute the SubDomainSet shapes
+        global_bounds = []
+        for i in self._global_bounds:
+            if isinstance(i, int):
```

Comment:
nitpicking: typically in Python we do this sort of things with a try-catch . "It's better to ask for forgiveness than for permission". And it's generally much faster

Reply:
A try/except catch doesn't work here since
`global_bounds.append(np.full(self._n_domains, i, dtype=np.int32))`
will not return an error even if `i` is already an array.